phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRepairRegionScanner.java

```diff
@@ -0,0 +1,426 @@
+        dataScan.setCacheBlocks(false);
+        try (ResultScanner resultScanner = dataHTable.getScanner(dataScan)) {
+            for (Result result = resultScanner.next(); (result != null); result = resultScanner.next()) {
+                ungroupedAggregateRegionObserver.checkForRegionClosing();
+                prepareExpectedIndexMutations(result, expectedIndexMutationMap);
+            }
+        } catch (Throwable t) {
+            ServerUtil.throwIOException(dataHTable.getName().toString(), t);
+        }
+        return expectedIndexMutationMap;
+    }
+
+    private Map<byte[], List<Mutation>> populateActualIndexMutationMap(Map<byte[], List<Mutation>> expectedIndexMutationMap) throws IOException {
+        Map<byte[], List<Mutation>> actualIndexMutationMap = Maps.newTreeMap(Bytes.BYTES_COMPARATOR);
+        Scan indexScan = prepareIndexScan(expectedIndexMutationMap);
+        try (RegionScanner regionScanner = region.getScanner(indexScan)) {
+            do {
+                ungroupedAggregateRegionObserver.checkForRegionClosing();
+                List<Cell> row = new ArrayList<Cell>();
+                hasMore = regionScanner.nextRaw(row);
+                if (!row.isEmpty()) {
+                    populateIndexMutationFromIndexRow(row, actualIndexMutationMap);
+                }
+            } while (hasMore);
+        } catch (Throwable t) {
+            ServerUtil.throwIOException(region.getRegionInfo().getRegionNameAsString(), t);
+        }
+        return actualIndexMutationMap;
+    }
+
+    private void repairAndOrVerifyIndexRows(Set<byte[]> dataRowKeys,
+                                            Map<byte[], List<Mutation>> actualIndexMutationMap,
+                                            IndexToolVerificationResult verificationResult) throws IOException {
+        List<Mutation> indexRowsToBeDeleted = new ArrayList<>();
+        Map<byte[], List<Mutation>> expectedIndexMutationMap = populateExpectedIndexMutationMap(dataRowKeys);
+        if (verifyType == IndexTool.IndexVerifyType.NONE) {
+            repairIndexRows(expectedIndexMutationMap, indexRowsToBeDeleted, verificationResult);
+            return;
+        }
+        if (verifyType == IndexTool.IndexVerifyType.ONLY) {
+            verifyIndexRows(actualIndexMutationMap, expectedIndexMutationMap, Collections.EMPTY_SET, Collections.EMPTY_LIST, verificationResult.getBefore(), true);
+            return;
+        }
+        if (verifyType == IndexTool.IndexVerifyType.BEFORE) {
+            verifyIndexRows(actualIndexMutationMap, expectedIndexMutationMap, Collections.EMPTY_SET, indexRowsToBeDeleted, verificationResult.getBefore(), true);
+            if (!expectedIndexMutationMap.isEmpty() || !indexRowsToBeDeleted.isEmpty()) {
+                repairIndexRows(expectedIndexMutationMap, indexRowsToBeDeleted, verificationResult);
+            }
+            return;
+        }
+        if (verifyType == IndexTool.IndexVerifyType.AFTER) {
```

Comment:
The AFTER option will not remove the extra verified rows in the index table. Same with the NONE option. Does it make sense to have these options when using the index table as the source ?

Reply:
I will leave this decision to be made within IndexTool. I think we can still allow these options and state in the help text for IndexTook that they do not remove the stale index rows.